### ISSUE TITLE
Variant of slice that doesn't perform clamp behavior

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#124572]
### 🐛 Describe the bug

When slicing into a tensor with data-dependent sizes, it is common to fail on these lines:

```
    if end_val < start_val:
        end_val = start_val
    elif end_val > sizes[dim]:
        end_val = sizes[dim]
```

The default behavior of slice is to clamp out of bounds start/end.

It would be useful to have a version of slice like https://gist.github.com/ezyang/0be810584a988ae5893961d77a3eacdd where the end/start are guaranteed to be in bounds, and so we can simply insert deferred runtime asserts testing in-boundedness

### Versions

main

**Comment by vadimkantorov** [comment]
If you do this, might also be good to make public the slice method in general :) https://github.com/pytorch/pytorch/issues/41625

**Comment by ezyang** [comment]
Any suggestions for the name? I'm going to do slice_strict without other opinions. cc @albanD

**Comment by vadimkantorov** [comment]
My problem with `narrow` in https://github.com/pytorch/pytorch/issues/41625 was that `narrow` did not support `step` argument

**Comment by ezyang** [comment]
Also, narrow is a differentiable decomposition to slice, so this does not actually work lol.

**Comment by albanD** [comment]
> My problem with narrow in https://github.com/pytorch/pytorch/issues/41625 was that narrow did not support step argument

Of course, but adding a new kwarg to an op is a lot simpler than adding a new op with unclear semantic.

> Also, narrow is a differentiable decomposition to slice, so this does not actually work lol.

Make narrow the leaf and make slice decompose to it? :D 


I am personally very much not happy with a public API that is `slice_strict`. It doesn't match any existing semantic (from other libs) and the "strict" is very much uninformative.

**Comment by ezyang** [comment]
> Make narrow the leaf and make slice decompose to it? :D

Is this a serious suggestion? Note that all of these changes if done in eager have big blast radius, so I have a strong incentive to only muck around with the decompositions when PT2 is turned on.

**Comment by albanD** [comment]
Ho I forgot we made narrow CompositeImplicit in eager as well. So yeah a bit disruptive there. Not sure about the impact on other decomps though.

But I really don't see how "slice_strict" makes sense as a public API.
I can't find any standard here from numpy/array API so maybe a less intrusive option is to have `slice(..., *, clamp=True)` ?

**Comment by ezyang** [comment]
@albanD this makes me afraid for different reasons haha, IIRC it's difficult to add new optional arguments to highly trafficked functions. But maybe I am misremembering.
